From 03656d6af21ab963e32b0c23c83455226afc546c Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Sat, 8 Feb 2025 16:17:36 +0800
Subject: [PATCH 51/98] desktop-shell: Add default window position via env

Support setting default window position via WESTON_DEFAULT_POSITION
environment variable. Accepts coordinates or predefined positions.

Tested on RK3588 EVB with:
WESTON_DEFAULT_POSITION=100,200 weston
WESTON_DEFAULT_POSITION=center weston
WESTON_DEFAULT_POSITION=left-top weston
WESTON_DEFAULT_POSITION=left-bottom weston
WESTON_DEFAULT_POSITION=right-top weston
WESTON_DEFAULT_POSITION=right-bottom weston

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 desktop-shell/shell.c | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/desktop-shell/shell.c b/desktop-shell/shell.c
index 5ee5b082..f3944761 100644
--- a/desktop-shell/shell.c
+++ b/desktop-shell/shell.c
@@ -3941,6 +3941,7 @@ weston_view_set_initial_position(struct weston_view *view,
 	struct weston_seat *seat;
 	pixman_rectangle32_t area;
 	struct weston_coord_global pos;
+	const char *buf;
 
 	if (view->has_position) {
 		pos.c = weston_coord(view->geometry.pos_offset.x,
@@ -3997,6 +3998,42 @@ weston_view_set_initial_position(struct weston_view *view,
 	if (range_y > 0)
 		y += random() % range_y;
 
+	buf = getenv("WESTON_DEFAULT_POSITION");
+	if (buf) {
+		int tmp_x, tmp_y;
+		if (sscanf(buf, "%d,%d", &tmp_x, &tmp_y) == 2) {
+			x = tmp_x;
+			y = tmp_y;
+		} else if (!strcmp(buf, "center")) {
+			x = area.x + range_x / 2;
+			y = area.y + range_y / 2;
+		} else if (!strcmp(buf, "center-left")) {
+			x = area.x;
+			y = area.y + range_y / 2;
+		} else if (!strcmp(buf, "center-right")) {
+			x = area.x + range_x;
+			y = area.y + range_y / 2;
+		} else if (!strcmp(buf, "center-top")) {
+			x = area.x + range_x / 2;
+			y = area.y;
+		} else if (!strcmp(buf, "center-bottom")) {
+			x = area.x + range_x / 2;
+			y = area.y + range_y;
+		} else if (!strcmp(buf, "left-top")) {
+			x = area.x;
+			y = area.y;
+		} else if (!strcmp(buf, "left-bottom")) {
+			x = area.x;
+			y = area.y + range_y;
+		} else if (!strcmp(buf, "right-top")) {
+			x = area.x + range_x;
+			y = area.y;
+		} else if (!strcmp(buf, "right-bottom")) {
+			x = area.x + range_x;
+			y = area.y + range_y;
+		}
+	}
+
 	pos.c = weston_coord(x, y);
 	weston_view_set_position(view, pos);
 }
-- 
2.39.5

